tsdb/querier.go

```diff
@@ -246,6 +246,11 @@ func PostingsForMatchers(ctx context.Context, ix IndexReader, ms ...*labels.Matc
 				return nil, err
 			}
 			its = append(its, allPostings)
+		case m.Type == labels.MatchRegexp && (m.Value == ".*" || m.Value == "^.*$"):
```

Comment:
I've added the regex (.*) and with anchor (^.*$) Because some test are directly with anchor as: https://github.com/prometheus/prometheus/blob/4b78238a30e88299bd3f8584869f5fa2a7084af3/tsdb/querier_test.go#L2806

And other ones (suggested ones as example) without anchor: 
https://github.com/prometheus/prometheus/blob/2beda6225cdac5452c5b18ca015462f6a814ba8c/tsdb/querier_test.go#L2961

Reply:
OK, the test failing is telling you something important.  This change without any change to `FastRegexMatcher` does not yield a valid implementation.

Always be suspicious of adding a special case to the implementation just to pass a test.

Reply:
I have some doubts here: AFAIK regex in prometheus are anchored: https://github.com/prometheus/prometheus/blob/91c05eed84c296089690cd0fbb7bac5976af6e53/model/labels/regexp.go#L72

Then (.*) and (^.*$) should be the same:
regexp has some differences:
```
"^(?:(?-s:.*))$"
"^(?:(?-ms:\\A.*$))$"
```
But string matcher is the same (as expected):
`labels.anyStringWithoutNewlineMatcher {}`

Then in test result of  (.*) and (^.*$) must be the same. As we are doing a fast path and not doing regex match for every value, it make sense to add both in the case:
`case m.Type == labels.MatchRegexp && (m.Value == ".*" || m.Value == "^.*$"):`

Also I can see different test with anchor and without.

Reply:
Maybe if I put it a different way: the change to `FastRegexMatcher` needs to be in this PR; having it in a separate PR #14540 does not yield a valid implementation.

How to check: when I change `TestPostingsForMatchers` to test a slightly different but equivalent regexp, the test needs to pass.

```
@@ -2803,7 +2803,7 @@ func TestPostingsForMatchers(t *testing.T) {
                {
-                       matchers: []*labels.Matcher{labels.MustNewMatcher(labels.MatchEqual, "n", "1"), labels.MustNewMatcher(labels.MatchRegexp, "i", "^.*$")},
+                       matchers: []*labels.Matcher{labels.MustNewMatcher(labels.MatchEqual, "n", "1"), labels.MustNewMatcher(labels.MatchRegexp, "i", "^.*")},
                        exp: []labels.Labels{
```

Reply:
Ok, got it. Thanks for the explanation. I've caught the issue:
This is not working correctly:
https://github.com/prometheus/prometheus/blob/0a68a9f420be325db6fd32b56eff78f92bd0e118/model/labels/regexp.go#L72

In fact this should fail, because the are variables with \n:
https://github.com/prometheus/prometheus/blob/0a68a9f420be325db6fd32b56eff78f92bd0e118/model/labels/regexp_test.go#L94

If you see regexp.Compile for example with ( .* ):
`m.re, err = regexp.Compile("^(?s:" + parsed.String() + ")$")`

Result of m.re is: `"^(?s:(?-s:.*))$"`
First ?s: allow dotall mode but second one ?-s: disables it, and this is added by `parsed.String() `

On the other hand you are right, this comparison:
`case m.Type == labels.MatchRegexp && (m.Value == ".*" || m.Value == "^.*$"):`
is wrong. Thinking in a better shortcut for .*

Reply:
A change like this one should work:
https://github.com/marioferh/prometheus/blob/9cfc5aad602bd96b6be84cb38b95851ea93d56b2/model/labels/regexp.go#L67

Reply:
I've pushed PR https://github.com/prometheus/prometheus/pull/14540 and a flag fix in regex.go#66

About (.*) Fastpath:
` "Why not ^.*, .*$, or any of the infinite equivalent expressions?"`

To implement a fastpath, the value needs to be compared to avoid regex matching. Common cases include (.*) and the anchored version like ^.*$. While we could technically add ^.* or .*$, I'm not sure how the comparison could be done differently in this context.  We can trim the anchor but maybe is not necessary and more error prone.

In fact the idea is: {n="1",i=~".*",j="foo} is equal to label matcher {n="1"j="foo}